File: core/tests/model_catalog.rs
use std::path::PathBuf;

use sona_core::models::catalog::{
    ModelListFilter, ModelSummary, list_models_with_installed_ids, select_models,
};
use sona_core::models::preset_models::LanguageMode;

fn model_summary(id: &str, languages: &[&str], installed: bool) -> ModelSummary {
    ModelSummary {
        id: id.to_string(),
        name: format!("{id} name"),
        model_type: "whisper".to_string(),
        languages: languages.iter().map(|code| code.to_string()).collect(),
        language_mode: LanguageMode::Selectable,
        size: "1 MB".to_string(),
        modes: vec!["batch".to_string()],
        installed,
        install_path: PathBuf::from(format!("C:/models/{id}")),
    }
}

#[test]
fn lists_models_with_injected_install_status() {
    let models_dir = PathBuf::from("C:/models");

    let models = list_models_with_installed_ids(
        &models_dir,
        &std::collections::HashSet::from(["sherpa-onnx-whisper-turbo".to_string()]),
    );

    assert!(models.iter().any(|model| {
        model.id == "sherpa-onnx-whisper-turbo"
            && model.installed
            && model.install_path == models_dir.join("sherpa-onnx-whisper-turbo")
    }));
    assert!(models.iter().any(|model| {
        model.id == "silero-vad"
            && !model.installed
            && model.install_path == models_dir.join("silero_vad.onnx")
    }));
}

#[test]
fn selects_models_by_mode_type_language_and_install_status() {
    let models = vec![
        ModelSummary {
            modes: vec!["batch".to_string()],
            ..model_summary("whisper-zh", &["en", "zh"], true)
        },
        ModelSummary {
            modes: vec!["streaming".to_string()],
            model_type: "zipformer".to_string(),
            language_mode: LanguageMode::Fixed,
            languages: vec!["zh".to_string()],
            installed: true,
            id: "stream-zh".to_string(),
            name: "stream-zh name".to_string(),
            size: "1 MB".to_string(),
            install_path: PathBuf::from("C:/models/stream-zh"),
        },
        ModelSummary {
            modes: vec!["batch".to_string()],
            model_type: "vad".to_string(),
            language_mode: LanguageMode::None,
            languages: vec![],
            installed: false,
            id: "vad-none".to_string(),
            name: "vad-none name".to_string(),
            size: "1 MB".to_string(),
            install_path: PathBuf::from("C:/models/vad-none"),
        },
    ];

    let selected = select_models(
        models,
        &ModelListFilter {
            mode: Some("batch".to_string()),
            model_type: Some("whisper".to_string()),
            language: Some("zh".to_string()),
            installed_only: true,
        },
    );

    assert_eq!(selected.len(), 1);
    assert_eq!(selected[0].id, "whisper-zh");
}

File: platforms/desktop/frontend/src/types/modelCatalog.ts
import presetModelsData from '../../../../../core/src/models/preset-models.json';
import type { ModelFileConfig } from './model';

export type TimestampSupportHint = 'token' | 'segment' | 'unknown';

/** How a model handles ASR language selection; drives client language pickers. */
export type LanguageMode = 'selectable' | 'auto' | 'fixed' | 'none';
export type ModelMode = 'live' | 'streaming' | 'batch';

export interface ModelRules {
  requiresVad: boolean;
  requiresPunctuation: boolean;
  timestampSupportHint?: TimestampSupportHint;
  initialRefreshRateMs?: number;
}

export interface ModelArtifact {
  url: string;
  filename: string;
  sha256?: string;
  sizeBytes?: number;
}

/**
 * Curated 1-5 quality hints rendered as dot meters on model cards. Values are
 * relative within the preset catalog, not benchmark scores, and are optional
 * so companion models can omit dimensions that make no sense for them.
 */
export interface ModelRatings {
  accuracy?: number;
  speed?: number;
}

/**
 * Curated family-level marks shown as tags and used by the label filter:
 * `accurate` for high-accuracy families, `lite` for lightweight ones.
 */
export type ModelLabel = 'accurate' | 'lite';

export interface ModelInfo {
  id: string;
  name: string;
  description: string;
  type:
    | 'zipformer'
    | 'sensevoice'
    | 'paraformer'
    | 'punctuation'
    | 'vad'
    | 'itn'
    | 'whisper'
    | 'funasr-nano'
    | 'fire-red-asr'
    | 'dolphin'
    | 'qwen3-asr'
    | 'parakeet-tdt'
    | 'moonshine'
    | 'speaker-segmentation'
    | 'speaker-embedding'
    | 'omnilingual';
  modes?: ModelMode[];
  /** All recognizable languages, sorted ascending ISO 639 codes (`yue` = Cantonese). */
  languages: string[];
  languageMode: LanguageMode;
  size: string;
  artifacts?: ModelArtifact[];
  isRecommended?: boolean;
  isArchive?: boolean;
  filename?: string;
  engine: 'sherpa-onnx' | 'llama-cpp';
  ratings?: ModelRatings;
  labels?: ModelLabel[];
  rules?: ModelRules;
  fileConfig?: ModelFileConfig;
  groupId?: string;
  versionLabel?: string;
}

export type ModelCatalogSectionType =
  | 'asr'
  | 'punctuation'
  | 'vad'
  | 'speaker-segmentation'
  | 'speaker-embedding';

export interface ModelCatalogModel extends ModelInfo {
  installPath: string;
  downloadPath: string;
  isInstalled: boolean;
  rules: ModelRules;
}

export interface ModelCatalogGroup {
  key: string;
  models: ModelCatalogModel[];
}

export interface ModelCatalogSection {
  type: ModelCatalogSectionType;
  groups: ModelCatalogGroup[];
}

export interface ModelSelectionOption {
  id: string;
  label: string;
  installPath: string;
  isInstalled: boolean;
}

export interface ModelCatalogSelectionOptions {
  streaming: ModelSelectionOption[];
  batch: ModelSelectionOption[];
  speakerSegmentation: ModelSelectionOption[];
  speakerEmbedding: ModelSelectionOption[];
}

export type ModelDependencyConfigKey = 'vadModelPath' | 'punctuationModelPath';

export interface ModelDependencyRequest {
  modelId: string;
  configKey: ModelDependencyConfigKey;
  installPath: string;
  isInstalled: boolean;
}

export interface ModelCatalogPathMatchToken {
  id: string;
  token: string;
}

export interface ModelCatalogRestoreDefaults {
  streamingModelPath?: string;
  batchModelPath?: string;
  vadModelPath?: string;
  punctuationModelPath?: string;
  speakerSegmentationModelPath?: string;
  speakerEmbeddingModelPath?: string;
  enableITN: boolean;
  batchVadEnabled?: boolean;
  vadBufferSize: number;
  maxConcurrent: number;
}

export interface ModelCatalogSnapshot {
  modelsDir: string;
  models: ModelCatalogModel[];
  sections: ModelCatalogSection[];
  selectionOptions: ModelCatalogSelectionOptions;
  modelPathById: Record<string, string>;
  modelIdByNormalizedPath: Record<string, string>;
  pathMatchTokens: ModelCatalogPathMatchToken[];
  dependencyRequestsByModelId: Record<string, ModelDependencyRequest[]>;
  restoreDefaults: ModelCatalogRestoreDefaults;
}

export interface ModelSelectionPaths {
  streamingModelPath: string;
  batchModelPath: string;
  speakerSegmentationModelPath: string;
  speakerEmbeddingModelPath: string;
}

export interface ModelCatalogSelectedIds {
  streaming: string | null;
  batch: string | null;
  speakerSegmentation: string | null;
  speakerEmbedding: string | null;
}

/** Selected preset-model ids for per-scenario (live/batch) companion models. */
export type ScenarioSelectedModelIds = Record<
  | 'liveSpeakerSegmentation'
  | 'batchSpeakerSegmentation'
  | 'liveSpeakerEmbedding'
  | 'batchSpeakerEmbedding'
  | 'livePunctuation'
  | 'batchPunctuation'
  | 'liveVad'
  | 'batchVad',
  string | null
>;

export const EMPTY_SCENARIO_SELECTED_MODEL_IDS: ScenarioSelectedModelIds = {
  liveSpeakerSegmentation: null,
  batchSpeakerSegmentation: null,
  liveSpeakerEmbedding: null,
  batchSpeakerEmbedding: null,
  livePunctuation: null,
  batchPunctuation: null,
  liveVad: null,
  batchVad: null,
};

export const DEFAULT_MODEL_RULES: ModelRules = {
  requiresVad: true,
  requiresPunctuation: false,
};

export const PRESET_MODELS: ModelInfo[] = presetModelsData as ModelInfo[];

export const PRESET_MODELS_MAP: Map<string, ModelInfo> = new Map(
  PRESET_MODELS.map((model) => [model.id, model])
);

/**
 * Snapshot models arrive from the Rust core without curated metadata, so
 * callers fall back to the shared preset JSON bundled with the frontend. Once
 * the core forwards these fields, the snapshot value wins.
 */
export function resolveModelRatings(model: ModelInfo): ModelRatings | undefined {
  return model.ratings ?? PRESET_MODELS_MAP.get(model.id)?.ratings;
}

/** Same fallback strategy as resolveModelRatings, for family labels. */
export function resolveModelLabels(model: ModelInfo): ModelLabel[] | undefined {
  return model.labels ?? PRESET_MODELS_MAP.get(model.id)?.labels;
}

export type ProgressCallback = (percentage: number, status: string, isFinished?: boolean) => void;
